### DISCUSSION TITLE
Example of running docker normally

### DISCUSSION THREAD

**Discussion by joserand** [Kitware/trame#378] [Q&A] [answered]
Hello, when I run the docker example,
![run](https://github.com/Kitware/trame/assets/89247520/740ec5ec-a2f7-4b76-ae5f-c3f407f90b9b)

 I want the image returned by the example to appear, as shown below

![example](https://github.com/Kitware/trame/assets/89247520/1279624e-bfab-4e0e-86f5-0413e5b5bba0)

But for the returned URL (http://localhost:9000/), it keeps loading, is my URL set incorrectly？ how should I get the URL I want to display?
Thanks for the reply！

**Comment by jourdain** [comment, marked as answer]
Because you start your docker on port 8888 (`-p 8888:80`), you should connect to `http://localhost:8888` not 9000.

But based on the error messages that you have, it seems that something else is wrong.

Could you provide more information on which example you are doing and on what system?

**Reply by joserand** [reply]
Thank you very much for your reply, I went into the http://localhost:8888 as suggested to browse, and as you said, I still can't get it running successfully.
I want to deploy a multi-user program on a server that can access services on my laptop when connected to the same network as the server. What happened next：
After downloading the code, I found that the pip download speed was slow, due to the regional problem, guessing that it was according to the requirmens .txt that there was no source change when downloading, and because I needed to create a python 3.8 virtual environment, but the actual virtual environment was not 3.8, I refactored the create_venv.sh and put the pip installation in it
![create](https://github.com/Kitware/trame/assets/89247520/15575882-c57f-480b-896a-9efb00a1a356)

When running the generated image, I got the following error, I found the corresponding error correction method from the Internet - refactored the apache2.conf file, and added ServerName localhost:80 to the end of the document
     Could not reliably determine the server's fully qualified domain name, using localhost,localdomain . Set the 'ServerName' directive globally to suppress this message.

My server system is CentOS-Linux, and the following is the configuration information that may be used
![system](https://github.com/Kitware/trame/assets/89247520/b5df3182-6c13-4d9d-9ef7-39acf1ceb051)
![docker](https://github.com/Kitware/trame/assets/89247520/d89f5397-1933-4c24-b555-77b1c3bb0411)

At the end, I have attached my revised code, which I hope will help you in your judgment
[SingleFile_try.zip](https://github.com/Kitware/trame/files/13454434/SingleFile_try.zip)

**Reply by joserand** [reply]
I found that after I didn't force the pytho 3.8 virtual environment, I successfully ran the given example on the http://localhost:8888/index.html you said.
But I still have a little problem, if I want to create a 3.8 environment, how should I change the code?The code I am currently create_env.sh is as follows, and the rest is unchanged as in the previous reply

    # # create virtual envir python3.8
    # pip install virtualenv -i http://pypi.douban.com/simple --trusted-host pypi.douban.com
    # cd ${TRAME_VENV%/*}
    # virtualenv  --python=3.8 ${TRAME_VENV##*/}
    
    python -m venv $TRAME_VENV
    . $TRAME_VENV/bin/activate
    
    # pip 
    pip install trame -i http://pypi.douban.com/simple --trusted-host pypi.douban.com
    pip install trame-vtk -i http://pypi.douban.com/simple --trusted-host pypi.douban.com
    pip install trame-vuetify -i http://mirrors.aliyun.com/pypi/simple/ --trusted-host mirrors.aliyun.com

**Comment by jourdain** [comment]
You need to change more files from the docker. In fact you should redo all the steps that we do in [trame](https://github.com/Kitware/trame/tree/master/docker) but with the proper [version](https://github.com/Kitware/trame/blob/master/docker/Dockerfile.pip#L6-L13) as your own docker file.

**Reply by joserand** [reply]
Can you give an example of the corresponding code modification. Because I think the virtual environment of 3.8 can be modified create_venv.sh, but I can't find an example of that on the web

**Reply by jourdain** [reply]
Just grab the docker directory of the trame repo and replace 3.9 by 3.8. Then rebuild the images for you.
